web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts

```diff
@@ -165,86 +171,97 @@ export default class OSKLayerGroup {
 
-    const x = coord.clientX;
+    for (let key of row.keys) {
+      const keySpec = key.spec;
+      if(keySpec.sp == ButtonClasses.blank || keySpec.sp == ButtonClasses.spacer) {
+        continue;
+      }
 
-    for (let k = 0; k < row.keys.length; k++) {
-      // Second-biggest, though documentation suggests this is probably right.
-      const keySquare = row.keys[k].square as HTMLElement; // gets the .kmw-key-square containing a key
-      const squareRect = keySquare.getBoundingClientRect();
+      // Max distance from the key's center to consider, horizontally.
+      const keyRadius = keySpec.proportionalWidth / 2;
+      const distanceFromCenter = Math.abs(proportionalCoords.x - keySpec.proportionalX);
 
       // Find the actual key element.
-      let childNode = keySquare.firstChild ? keySquare.firstChild as HTMLElement : keySquare;
-
-      if (childNode.className !== undefined
-        && (childNode.className.indexOf('key-hidden') >= 0
-          || childNode.className.indexOf('key-blank') >= 0)) {
-        continue;
-      }
-      const x1 = squareRect.left;
-      const x2 = squareRect.right;
-      if (x >= x1 && x <= x2) {
-        // Within the key square
-        return <KeyElement>childNode;
-      }
-      dx = x1 - x;
-      if (dx >= 0 && dx < dxMin) {
-        // To right of key
-        closestKeyIndex = k; dxMin = dx;
-      }
-      dx = x - x2;
-      if (dx >= 0 && dx < dxMin) {
-        // To left of key
-        closestKeyIndex = k; dxMin = dx;
+      if(distanceFromCenter - keyRadius <= 0) {
+        // As noted above:  if we land within a key's square, match instantly!
+        return key.btn;
+      } else {
+        const distance = distanceFromCenter - keyRadius;
+        if(distance < minDistance) {
+          minDistance = distance;
+          closestKey = key;
+        }
       }
     }
 
-    if (dxMin < 100000) {
-      const t = <HTMLElement>row.keys[closestKeyIndex].square;
-      const squareRect = t.getBoundingClientRect();
+    // Step 3:  If the input coordinate wasn't within any valid key's "square",
+    // determine if the nearest valid key is acceptable.
+    const minHorizFudge = NEAREST_KEY_MIN_HORIZONTAL_FUDGE_PIXELS / this.element.offsetWidth;
 
-      const x1 = squareRect.left;
-      const x2 = squareRect.right;
-
-      // Limit extended touch area to the larger of 0.6 of key width and 24 px
-      if (squareRect.width > 40) {
-        dxMax = 0.6 * squareRect.width;
-      }
+    // If the condition is not met, there are no valid keys within this row.
+    if (minDistance < Number.MAX_VALUE) {
+      let fudgeFactor = closestKey.spec.proportionalWidth * NEAREST_KEY_HORIZ_FUDGE_FACTOR;
+      fudgeFactor = fudgeFactor > minHorizFudge ? fudgeFactor : minHorizFudge;
 
-      if (((x1 - x) >= 0 && (x1 - x) < dxMax) || ((x - x2) >= 0 && (x - x2) < dxMax)) {
-        return <KeyElement>t.firstChild;
+      if(minDistance <= fudgeFactor) {
+        return closestKey.btn;
       }
     }
```

Comment:
Shouldn't the acceptable area be a fixed value, not relative to the size of the key? That is, pressing on the edge of spacebar shouldn't have a different outcome to pressing the edge of the Z key in terms of how close the finger has to be.

Reply:
I'm just working with "what was" here.  What you're saying certainly makes sense, though.

Given most of our keyboards tend to have 10 keys in their top row, using its full width... that implies a default total key width + padding of 10% `computedWidth`.  (Thus, 0.1 "proportional width + padding").  Would a 6% * `computedWidth` fudge factor be reasonable?

Reply:
yep, sounds like we came to the same conclusion 🤣 in my previous comment https://github.com/keymanapp/keyman/pull/11129#discussion_r1548878929

So that's a good sign!